src/distribute_kv_cache/rpc/utils.rs

```diff
@@ -0,0 +1,123 @@
+use bytes::{Buf, BytesMut};
+use tracing::debug;
+
+use super::error::RpcError;
+
+/// Read with timeout options from the environment variables
+#[macro_export]
+macro_rules! read_exact_timeout {
+    ($self:expr, $dst:expr, $read_timeout:expr) => {
+        async move {
+            use tokio::time::timeout;
+
+            match timeout($read_timeout, $self.read_exact($dst)).await {
+                Ok(res) => match res {
+                    Ok(size) => Ok(size),
+                    Err(read_err) => Err(read_err),
+                },
+                Err(timeout_err) => Err(timeout_err.into()),
+            }
+        }
+    };
+}
+
+/// Write with timeout options from the environment variables
+#[macro_export]
+macro_rules! write_all_timeout {
+    ($self:expr, $src:expr, $write_timeout:expr) => {
+        async move {
+            use tokio::time::timeout;
+
+            match timeout($write_timeout, $self.write_all($src)).await {
+                Ok(res) => match res {
+                    Ok(size) => Ok(size),
+                    Err(write_err) => Err(write_err),
+                },
+                Err(timeout_err) => Err(timeout_err.into()),
+            }
+        }
+    };
+}
+
+/// Write vectored with timeout options from the environment variables
+#[macro_export]
+macro_rules! write_vectored_timeout {
+    ($self:expr, $src:expr, $write_timeout:expr) => {
+        async move {
+            use tokio::time::timeout;
+
+            // TODO: check all the write_vectored is done
+            match timeout($write_timeout, $self.write_vectored($src)).await {
+                Ok(res) => match res {
+                    Ok(size) => Ok(size),
+                    Err(write_err) => Err(write_err),
+                },
+                Err(timeout_err) => Err(timeout_err.into()),
+            }
+        }
+    };
+}
+
+/// Connect a stream with timeout options from the environment variables
+#[macro_export]
+macro_rules! connect_timeout {
+    ($addr:expr, $connect_timeout:expr) => {
+        async move {
+            use tokio::net::TcpStream;
+            use tokio::time::timeout;
+
+            match timeout($connect_timeout, TcpStream::connect($addr)).await {
+                Ok(res) => match res {
+                    Ok(stream) => Ok(stream),
+                    Err(connect_err) => Err(connect_err),
+                },
+                Err(timeout_err) => Err(timeout_err.into()),
+            }
+        }
+    };
+}
+
+/// Try to get u64 data from buffer with range
+pub fn get_u64_from_buf(buf: &[u8], start: usize) -> Result<u64, RpcError> {
+    buf.get(start..start + 8)
+        .ok_or_else(|| RpcError::InternalError("Invalid range data".to_owned()))
```

Comment:
This message can be more meaningful if you put the start and the length of buf into the message.

Reply:
Got, I have updated with a formatted error message.

Reply:
BTW, this function will be removed because I have use bincode as the deserializer.